### PR TITLE
Support for OpenAPIv3

### PR DESCRIPTION
**Pull request by HEM42**
Due to the structure change in OpenAPI v3, body parameters has been moved to 'RequestBody'

This PR handles this fact, and brings the missing OpenAPIv3 support

### CODE REVIEW

File: lib/Mojolicious/Command/openapi.pm [hunk 1]
```diff
@@ -41,6 +41,7 @@ sub run {
     'p|parameter=s'          => \my %parameters,
     'c|content=s'            => \my $content,
     'S|response-size=i'      => sub { $ua{max_response_size} = $_[1] },
+    's|schema-version=s'     => \my $schema,
```
Comment (jhthorsen) [inline lib/Mojolicious/Command/openapi.pm:44]: Maybe this could just be "schema" instead of "schema-version"?

File: lib/OpenAPI/Client.pm [hunk 2]
```diff
@@ -19,8 +19,19 @@ has base_url => sub {
   my $schema  = $self->validator->schema;
   my $schemes = $schema->get('/schemes') || [];
 
-  return Mojo::URL->new->host_port($schema->get('/host') || 'localhost')->path($schema->get('/basePath') || '/')
-    ->scheme($schemes->[0] || 'http');
+  my $url;
+
+  # Schema loaded is an OpenAPIv3 spec
+  if ($schema->get('/openapi')) {
+    $url = Mojo::URL->new($schema->get('/servers')->[0]{url});
```
Comment (jhthorsen) [inline lib/OpenAPI/Client.pm:26]: What happens if there are no servers defined?

File: lib/OpenAPI/Client.pm [hunk 3]
```diff
@@ -188,6 +199,30 @@ sub _build_tx {
     }
   }
 
+  # Handling requestBody for OpenAPIv3
+  if ($op_spec->{requestBody}) {
+    my $content_key = [keys %{$op_spec->{requestBody}{content}}]->[0];
+    my $ct          = {'application/x-www-form-urlencoded' => 'form', 'application/json' => 'json'}->{$content_key};
```
Comment (jhthorsen) [inline lib/OpenAPI/Client.pm:205]: I'm not sure if I like this line two much. Could you use a ternary operator instead? And maybe default to "application/json"?

File: lib/Mojolicious/Command/openapi.pm (hunk 1 above)
Comment (HEM42) [reply lib/Mojolicious/Command/openapi.pm:44]: Addressing this in the next push

File: lib/OpenAPI/Client.pm (hunk 2 above)
Comment (HEM42) [reply lib/OpenAPI/Client.pm:26]: You are right, according to the spec, servers could be omitted from the spec or be an empty array. If this is the case, it should default to '/'.

Addressing this in the next push

File: lib/OpenAPI/Client.pm (hunk 3 above)
Comment (HEM42) [reply lib/OpenAPI/Client.pm:205]: Addressing this in the next push

**Comment by HEM42** [comment]
> I'm very positive to adding support for v3, but I wish I had finished [mojolicious/json-validator#189](https://github.com/mojolicious/json-validator/pull/189) because I think that would help this module as well...

I undertand, I have added a few comments to you PR.

Looking at the upcoming changes to J::V, it seems straight forward to support here as well

**Comment by pplu** [comment]
Now that OpenAPI supports v3, is there any possiblity of getting v3 support merged into OpenAPI::Client? I'm currently building a v3 API with Mojolicious::Plugin::OpenAPI, and this would be great for building tests :)

**Comment by jhthorsen** [comment]
This should be fixed in v1.00

**Comment by pplu** [comment]
Hey! Thanks for releasing support for Openapi v3!

**Pull request closed without merging**